packages/x-date-pickers/src/internals/hooks/useField/useFieldCharacterEditing.ts

```diff
@@ -0,0 +1,442 @@
+
+/**
+ * Update the active section value when the user pressed a key that is not a navigation key (arrow key for example).
+ * This hook has two main editing behaviors
+ *
+ * 1. The numeric editing when the user presses a digit
+ * 2. The letter editing when the user presses another key
+ */
+export const useFieldCharacterEditing = <TDate, TSection extends FieldSection>({
+  sections,
+  updateSectionValue,
+}: UseFieldEditingParams<TDate, TSection>) => {
+  const utils = useUtils<TDate>();
+
+  const [query, setQuery] = React.useState<CharacterEditingQuery | null>(null);
+
+  React.useEffect(() => {
+    if (query != null && sections[query.sectionIndex]?.dateSectionName !== query.dateSectionName) {
+      setQuery(null);
+    }
+  }, [sections, query]);
+
+  React.useEffect(() => {
+    if (query != null) {
+      const timeout = setTimeout(() => setQuery(null), QUERY_LIFE_DURATION_MS);
+
+      return () => {
+        window.clearTimeout(timeout);
+      };
+    }
+
+    return () => {};
+  }, [query]);
+
+  const applyQuery = (
+    { keyPressed, sectionIndex }: ApplyCharacterEditingParams,
+    getFirstSectionValueMatchingWithQuery: QueryApplier<TSection>,
+  ): ReturnType<CharacterEditingApplier<TDate, TSection>> => {
+    const cleanKeyPressed = keyPressed.toLowerCase();
+    const activeSection = sections[sectionIndex];
+
+    // The current query targets the section being editing
+    // We can try to concatenated value
+    if (query != null && query.sectionIndex === sectionIndex) {
+      const concatenatedQueryValue = `${query.value}${cleanKeyPressed}`;
+
+      const queryResponse = getFirstSectionValueMatchingWithQuery(
+        concatenatedQueryValue,
+        activeSection,
+      );
+      if (!isQueryResponseWithoutValue(queryResponse)) {
+        setQuery({
+          sectionIndex,
+          value: concatenatedQueryValue,
+          dateSectionName: activeSection.dateSectionName,
+        });
+        return queryResponse;
+      }
+    }
+
+    const queryResponse = getFirstSectionValueMatchingWithQuery(cleanKeyPressed, activeSection);
+    if (isQueryResponseWithoutValue(queryResponse) && !queryResponse.saveQuery) {
+      setQuery(null);
+    } else {
+      setQuery({
+        sectionIndex,
+        value: cleanKeyPressed,
+        dateSectionName: activeSection.dateSectionName,
+      });
+    }
+
+    if (isQueryResponseWithoutValue(queryResponse)) {
+      return null;
+    }
+    return queryResponse;
+  };
+
+  const applyLetterEditing: CharacterEditingApplier<TDate, TSection> = (params) => {
+    const getFirstSectionValueMatchingWithQuery: QueryApplier<TSection> = (
+      queryValue,
+      activeSection,
+    ) => {
+      switch (activeSection.dateSectionName) {
+        case 'month': {
+          const getMonthResponse = (format: string): ReturnType<QueryApplier<TSection>> => {
+            const matchingMonths = utils
+              .getMonthArray(utils.date()!)
```

Comment:
Nitpick, but this could probably be memoized somewhere as its result should not change, unless locale changes. 🤔

Reply:
We can do a follow up where we put this info in the `boundaries` (and rename it of course).